### ISSUE TITLE
Crash on startup

### ISSUE THREAD

**Issue by marcust** [plietar/librespot#55]
Hey, as one is invited to experiment with it I did so today, with mixed results. But maybe that's due to the fact that I'm trying the crazy thing on a Raspberry Pi (I want a spotify connect target on a Pi).

It took me just two days to get a working rust on the Pi but finally I was able to build librespot. My startup looks like this:

```
target/release/librespot --appkey ../spotify_appkey.key --username 1127156859 --password ******** --cache .tmp --name playpi
Authenticated !
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.front
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround21
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround21
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround40
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround41
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround50
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround51
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.surround71
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.iec958
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.iec958
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.iec958
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.hdmi
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.hdmi
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.modem
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.modem
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.phoneline
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.phoneline
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
kMessageTypeHello playpi 61672252dc61a1a7a07debc7ed0d17944194ceee 1 1457222331920
kMessageTypeNotify eeehal fbffcd40ab9c48f2fbf091733a0f55a8e3b3866f 6 1457203530953
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: WireError("invalid value for enum: 10")', ../src/libcore/result.rs:746
note: Run with `RUST_BACKTRACE=1` for a backtrace.
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', ../src/libcore/result.rs:746
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: "SendError(..)"', ../src/libcore/result.rs:746
```

Then the process terminates. I can see the playpi target in a Spotify client for a short time. Neither linux audio nor rust are any of my strong sides but if I can be of any assistance here I'll try to be helpfull.

The output opens a couple of questions:
1. are unknown PCMs a bad thing?
2. what is this server socket he is trying to connect to? Is that a bad thing?
3. what is this server request channel? is it a bad thing if it can not connect?
4. I figured what jack is, but apparently I'm unable to run it (yet). But is it needed anyway?

**Comment by plietar** [comment]
I've just pushed a fix for this (5059432). Spotify have made a tiny enhancement to the protocol which librespot didn't understand.
I'm not sure about the ALSA/Jack warnings, but it should be fine.

**Comment by herrernst** [comment]
I also always get the "Unknown PCM" and "Cannot connect to server" messages, but I ignore them and it works fine.

**Comment by marcust** [comment]
@plietar cool, works for me now

@herrernst I uninstalled a couple of audio related packages, first of all jackd packages and now the cannot connect and jack messages are gone.

**Comment by arigit** [comment]
@marcust what packages did you remove to get the jackd and "connect" errors gone? 
I am running OSMC and I couldn't find  which package to uninstall (jackd / jackd2 / jack-tools etc are not installed).
While you were getting the jack errors, was librespot discovered by the clients and was spotify-connect working for you?

librespot doesn't crash for me anymore (using the latest  build from [herrernst](https://github.com/plietar/librespot/issues/56)), however the spotify clients do not "see" the raspberry (I test with Android and Linux spotify clients). 
I am using a HiFiBerry DAC+ on OSMC. Works fine with Kodi, and it worked just once with librespot, not sure what changed but after lots of trial and error, I can't get it to work again.

---

...
ALSA lib pcm.c:2239:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.modem
ALSA lib pcm.c:2239:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.phoneline
ALSA lib pcm.c:2239:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.phoneline
ALSA lib pcm_direct.c:896:(snd1_pcm_direct_initialize_slave) slave plugin does not support mmap interleaved or mmap noninterleaved access
ALSA lib pcm_dmix.c:1034:(snd_pcm_dmix_open) unable to initialize slave
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
kMessageTypeHello Raspberry b7870749a82019bec25fcc2db279951e62cdbd42 1 1457712842412
kMessageTypeHello Galaxy S6 b78f746216845f1309d617a88abc6e62c923d1ec 48 0
kMessageTypeNotify Galaxy S6 b78f746216845f1309d617a88abc6e62c923d1ec 49 0
...

---

$ aplay -l
***\* List of PLAYBACK Hardware Devices ****
card 0: sndrpihifiberry [snd_rpi_hifiberry_dacplus], device 0: HiFiBerry DAC+ Pro HiFi pcm512x-hifi-0 []
  Subdevices: 0/1
  Subdevice #0: subdevice #0

**Comment by marcust** [comment]
@arigit according to my apt history I uninstalled the following packages:
pulseaudio jackd jackd2 libjack-jackd2-0 
and some autoremoves that removed depending packages. Then I got compile error and reinstalled 
libportaudio2 portaudio19-dev

That made the warnings go away.

While I was having those errors it was discoverable for a short while before it crashed, this was due to the original bug I figure.

It has worked for me now ever since, though sometime it is not discoverable till I restart it.
